### PR TITLE
COSI-72: Unit Tests Improvements

### PR DESCRIPTION
**Pull request by anurag4DSB**
### Improvements to Tests and Code Coverage

This pull request focuses on enhancing the overall testing structure, improving code readability, and increasing test coverage from 78% to 90%. Key changes include:
- **Common Mocks for S3 and IAM:** Centralized mock implementations for S3 and IAM services, reducing code duplication and improving maintainability.
- **Enhanced IAM Service Testing:** Instead of mocking the CreateBucketAccess helper, failures are now simulated using the actual IAM service, providing more realistic failure scenarios.
- **Adoption of Ginkgo SpecContext:** Replaced older constructs with SpecContext where applicable, aligning tests with Ginkgo best practices.
- **Removal of Redundant Comments:** Removed unnecessary comments in test cases to improve readability and focus on essential details.
- **Refactored and Expanded Tests:** Refactored existing test cases for better logical grouping and added new tests to cover edge cases and logically possible scenarios.
- **Improved Error Handling:** Enhanced error handling during initialization, making the codebase more robust and easier to test.
- **Reusable Mock Packages:** Replaced inline mocks with reusable packages to streamline future test development.

### CODE REVIEW

**Review by dvasilas — APPROVED** [review]

File: pkg/grpcfactory/interceptors_test.go [hunk 1]
```diff
@@ -39,7 +37,7 @@ var _ = Describe("gRPC Factory Interceptors", func() {
 			Expect(err).NotTo(HaveOccurred())
 		})
 
-		It("should handle invocation error and log it", func() {
+		It("should handle invocation error and log it", func(ctx SpecContext) {
```
Comment (fredmnl) [inline pkg/grpcfactory/interceptors_test.go:40]: For my own understanding, what is adding the `ctx` here doing? Is it just for consistency?

File: pkg/grpcfactory/server_test.go [hunk 2]
```diff
@@ -141,5 +141,26 @@ var _ = Describe("gRPC Factory Server", func() {
...
 		})
+
+		It("should return an error for unsupported address schemes", func(ctx SpecContext) {
+			invalidAddress := "http://invalid-scheme-address" // Address with an unsupported scheme
+
+			server, err := grpcfactory.NewCOSIProvisionerServer(invalidAddress, identityServer, provisionerServer, nil)
+			Expect(err).NotTo(HaveOccurred()) // Ensure server creation succeeds
+			Expect(server).NotTo(BeNil())
+
+			errChan := make(chan error, 1)
```
Comment (fredmnl) [inline pkg/grpcfactory/server_test.go:152]: Wondering if we can just timeout this entire test case to simplify the way it's written. Perusing the gingko doc, I think `SpecTimeout` might do the trick but I'm not sure. If it applies, we could just synchronously wait for `server.Run` to return an error and check it. If it does not, the test will time out which is fine.

**Review by fredmnl — APPROVED** [review]

File: pkg/grpcfactory/interceptors_test.go (hunk 1 above)
Comment (anurag4DSB) [reply pkg/grpcfactory/interceptors_test.go:40]: Not needed indeed.
It was a find and replace 😓

Comment (anurag4DSB) [reply pkg/grpcfactory/interceptors_test.go:40]: My bad, I thought it was for another test, github hid the usage
I thought the usage is not there but I am actually using it.
Yes I am using it for `err := grpcfactory.ApiLogger(ctx, method, req, reply, cc, invoker)`

Adding ctx here primarily ensures that the test works within Ginkgo’s managed context lifecycle. This allows the invoker to receive a properly managed context.Context, which supports cancellation and timeout handling. Even if this particular test doesn’t leverage these features, including ctx ensures consistency with real-world function signatures and enables future flexibility (e.g., if the invoker uses context-sensitive logic like deadlines or tracing).

Yes, I agree, I actually started `specContext` following this after the comment from @jonathan-gramain in another PR. But using a context helps anyways, whether its `context.TODO` or `context.Background`

Comment (anurag4DSB) [reply pkg/grpcfactory/interceptors_test.go:40]: that said we do need a styling guide for GO indeed

File: pkg/grpcfactory/server_test.go (hunk 2 above)
Comment (anurag4DSB) [reply pkg/grpcfactory/server_test.go]: Thats a great idea
Added an extra commit https://github.com/scality/cosi-driver/pull/56/commits/495c5ff1a128284cd4d52f664e96467ce91bb8ec
